Make metrics collection optional/faster

QuerthDP opened the pull request:
This patch contains:
- Implementation of lock-free histogram with hot and cold bucket pools
- Introduction of crate feature "metrics" to make using them optional
- Functionality to gather latency statistics via histogram snapshots
- Implementation of [rates of queries per second](https://github.com/scylladb/cpp-driver/blob/9d6b05c9d4ebd0a6d7006af4df3e33fcdf956eeb/src/metrics.hpp#L39C1-L252C5) based on cpp-driver implementation.
- Initial implementation of connection metrics (changes required after review)

Fixes: #330

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

QuerthDP commented:
Changelog:

- rebased on main

muzarski commented on scylla/src/observability/metrics.rs:
```diff
@@ -1,24 +1,19 @@
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "metrics error: {}", self.cause)
-    }
+#[derive(Error, Debug, PartialEq)]
+pub enum MetricsError {
+    #[error("lock-free histogram error")]
+    LFHistogramErr(#[from] LFError),
+    #[error("histogram error")]
+    HistogramErr(#[from] histogram::Error),
+    #[error("histogram is empty")]
+    Empty,
```
I know that previous `MetricsError` was not documented, but we will appreciate this change as a bonus :). This will make a progress towards https://github.com/scylladb/scylla-rust-driver/issues/519

muzarski commented on scylla/src/observability/metrics.rs:
```diff
@@ -120,4 +118,24 @@ impl Metrics {
+
+    // Metric implementations
+
+    fn mean(h: Histogram) -> Result<u64, MetricsError> {
```
I wonder why they decided to remove `Histogram::mean()` method

brayniac replied:
The histogram has no way of providing a true mean. Do we use the lower or upper end of the bucket range while calculating? Somewhere in the middle? It seems more appropriate to let the caller decide how they want to deal with this detail. Same when determining a percentile, the best we can do is return the `Bucket` where the percentile falls into its range. It may depend on your use-case on what value to report. Previous assumptions of over-reporting latencies by using the upper-edge of the bucket might not be appropriate for all use-cases.

muzarski commented on examples/Cargo.toml:
```diff
@@ -18,6 +18,7 @@ scylla = { path = "../scylla", features = [
     "num-bigint-03",
     "num-bigint-04",
     "bigdecimal-04",
+    "metrics",
```
Please, add a CI step with cargo checks with this feature enabled. You can see examples for other features in `.github/workflows/rust.yml` (`build` job). It can be added to this, or a following commit.

muzarski reviewed (commented):
I still need to review commits that introduce `Meter` and connection metrics. Posting this review early, because there are some matters to discuss.

NikodemGapski commented on scylla/src/observability/lock_free_histogram.rs:
```diff
@@ -0,0 +1,303 @@
+    }
+}
+
+impl Default for LockFreeHistogram {
+    fn default() -> Self {
+        // Config: 64ms error, values in range [0ms, ~262_000ms].
+        // Size: (2^13 + 5 * 2^12) * 8B * 2 ~= 450kB.
+        let grouping_power = 12;
+        let max_value_power = 18;
+        LockFreeHistogram::new(grouping_power, max_value_power)
+    }
+}
```
Since the histogram crate no longer provides defaults, I had to come up with some choice here. It was my best guess at what might be needed, though that is obviously to be discussed and modified if needed.

brayniac replied:
You may find our calculator useful while evaluating what parameters are appropriate: https://observablehq.com/@iopsystems/h2-histogram

wprzytula replied:
:question: @NikodemGapski Have you consulted the above calculator about the defaults?

NikodemGapski replied:
I mean, I had *made* my guess according to the mentioned calculator, but I can't know if it meets the needs of the users of this driver (the range and absolute error I noted in the comment for reference). It is your decision whether or not to change it.

muzarski commented on scylla/src/observability/lock_free_histogram.rs:
```diff
@@ -0,0 +1,303 @@
+    Mutex,
+}
+
+pub struct LockFreeHistogram {
```
Thank you for the explanation! I think part of this deserves to be put in the docstring.

> Also, should you choose to go with `AtomicHistogram`, the change will be rather effortless, as I maintained the API schema used in crate histogram for my implementation.

This is great. Also, please unpub (`pub(crate)`) the `LockFreeHistogram` and its methods. Only then will we be 100% sure that if we ever decide to drop/modify `LockFreeHistogram`, such change will not be API breaking.

Which histogram implementation we want to incorporate to the driver, well this needs to be discussed. I'd wait until @Lorak-mmk and @wprzytula review the code.

NikodemGapski replied:
I unpubbed `LockFreeHistogram` and its methods, but now I can see it might be difficult to make such a change completely non-API-breaking. That's because `LockFreeHistogramError` is propagated as `MetricsError`'s cause and thus needs to be pub. Upon change to `AtomicHistogram` this error struct would be removed entirely.

I'm not yet sure how we could go around this issue.

muzarski replied:
Ah, good catch. There are some workarounds for this, however. For example, we could always hide underlying `LockFreeHistogramError` under `Arc<dyn Error>`. I wouldn't worry about this now. Let's wait for others to join the discussion.

brayniac replied:
For what it's worth, while `.load()` is not atomic wrt concurrent increments into the histogram. I'd still consider using the `AtomicHistogram` which is used in metrics for both [rpc-perf ](https://github.com/iopsystems/rpc-perf) and [Pelikan](https://github.com/pelikan-io/pelikan).

As I said in the issue opened in rustcommon, I'm not sure that the potential skew here would be meaningful. But I do welcome some concrete details if such skew does prove to be meaningful.

The TLDR is I'm not sure how much it matters whether it's on one side or the other of loading a histogram. If you envision periodically snapshotting the histogram, it seems you have to accept that the latency is already being recorded at the tail end of the event. Imagine a request that takes a long time, that latency gets incremented after the fact, when the service might be back to responding quickly. 

My feeling is that ultimately this is all an approximation and I've found the `AtomicHistogram` as in the `histogram` crate to be satisfactory for projects I work on.

brayniac commented on scylla/src/observability/metrics.rs:
```diff
@@ -138,4 +181,58 @@ impl Metrics {
+            .map(|bucket| bucket.count() as u128)
+            .sum::<u128>();
+
+        let mean = Self::mean(h.clone())? as u128;
+        let mut variance_sum = 0;
+        for bucket in h.into_iter() {
+            let count = bucket.count() as u128;
+            let mid = ((bucket.start() + bucket.end()) / 2) as u128;
+
+            variance_sum += count * (mid as i128 - mean as i128).pow(2) as u128;
+        }
+        let variance = variance_sum / total_count;
+
+        Ok((variance as f64).sqrt() as u64)
+    }
+
+    fn minmax(h: Histogram) -> Result<(u64, u64), MetricsError> {
+        let mut min = u64::MAX;
+        let mut max = 0;
+        for bucket in h.into_iter() {
+            if bucket.count() == 0 {
+                continue;
+            }
+            let lower_bound = bucket.start();
+            let upper_bound = bucket.end();
+
+            if lower_bound < min {
+                min = lower_bound;
+            }
+            if upper_bound > max {
+                max = upper_bound;
+            }
+        }
+
+        if min > max {
+            Err(MetricsError::Empty)
+        } else {
+            Ok((min, max))
+        }
+    }
```
I believe that `IntoIterator` is needed for the for loop syntax to work so here you could do:

```rust
for bucket in &h {
        // some code here
}
```

But there's no reason not to have `.iter()` as well, so I've added that in `histogram-0.11.2` which is now live on crates.io

wprzytula replied:
> I believe that `IntoIterator` is needed for the for loop syntax to work so here you could do:
> 
> ```rust
> for bucket in &h {
>         // some code here
> }
> ```

You're right, actually [the standard library defines `IntoIterator` on references to collections as well](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#implementors). TIL, thank you! :bow:  

> But there's no reason not to have `.iter()` as well, so I've added that in `histogram-0.11.2` which is now live on crates.io

Great! :rocket:

QuerthDP replied:
Indeed there was no need of cloning the histogram 😅 Rookie mistake.
Works fine as intended with the explanation and code snippets provided by you! Thanks! 🙇

NikodemGapski commented on scylla/src/observability/lock_free_histogram.rs:
```diff
@@ -0,0 +1,303 @@
+        // The number of finished snapshot *threads*.
+        let finished_snapshots = Arc::new(AtomicU64::new(0));
+
+        // Observer thread lambda.
+        let h_1 = h.clone();
+        let fin_s_1 = finished_snapshots.clone();
+        let observer = move || {
+            while fin_s_1.load(Ordering::Relaxed) < snapshot_count {
+                h_1.increment(0).unwrap();
+            }
+        };
+
+        // Snapshot thread lambda.
```
Oh, not at all :laughing: . I just used it as a synonym for a closure. I can change it to `closure` if you believe it will improve understandability. :+1:

wprzytula replied:
:smile: Yes, please. Apparently the Rust world tries so hard to cut itself off C++ that I didn't think of it.

QuerthDP commented on scylla/src/observability/metrics.rs:
```diff
@@ -223,6 +229,26 @@ impl Metrics {
+        self.connection_timeouts.fetch_add(1, ORDER_TYPE);
+    }
+
+    /// Increments counter for request timeouts
+    pub(crate) fn inc_request_timeouts(&self) {
+        self.request_timeouts.fetch_add(1, ORDER_TYPE);
+    }
```
AFAIK this metric was used in cpp-driver to count client timeouts on requests to the database.
So is this method called in our implementation (alongside returning QueryError::RequestTimeout, the docstring of which mentions that it regards the client side).

I think it should be specified in the docstring of our inc method as well, so I will adjust it.

QuerthDP commented on scylla/src/observability/metrics.rs:
```diff
@@ -223,6 +229,26 @@ impl Metrics {
         self.retries_num.fetch_add(1, ORDER_TYPE);
     }
 
+    /// Increments counter for total number of connections
+    pub(crate) fn inc_total_connections(&self) {
+        self.total_connections.fetch_add(1, ORDER_TYPE);
+    }
+
+    /// Decrements counter for total number of connections
+    pub(crate) fn dec_total_connections(&self) {
+        self.total_connections.fetch_sub(1, ORDER_TYPE);
+    }
```
Those functions are not meant to be called by the user.

Nevertheless I surely will adjust the docstrings to better describe the logic behind those functions for devs to understand and use properly if needed in other places in the future.

wprzytula replied:
> for devs to understand and use properly if needed in other places in the future.

That's what I had in mind - internal usage.

QuerthDP commented on scylla/src/observability/metrics.rs:
```diff
@@ -16,6 +16,23 @@ pub enum MetricsError {
+    pub percentile_98: u64,
+    pub percentile_99: u64,
+    pub percentile_99_9: u64,
+}
```
So far I've used the `#[non_exhaustive]` macro, but this case is still open for discussion.

QuerthDP commented:
Changelog:
- rebased on main

wprzytula commented on scylla/src/network/connection_pool.rs:
```diff
@@ -61,6 +64,12 @@ pub(crate) struct PoolConfig {
     pub(crate) pool_size: PoolSize,
     pub(crate) can_use_shard_aware_port: bool,
     pub(crate) keepalive_interval: Option<Duration>,
+    #[cfg(feature = "metrics")]
+    // TODO: The metrics should definitely not be stored here,
+    //       but it was the easiest way to pass it to the refiller.
+    //       It could be refactored to be passed as a parameter to the refiller,
+    //       but it would require a lot of changes in the code.
+    pub(crate) metrics: Option<Arc<Metrics>>,
```
:wrench: I refactored passing `Metrics` through the code. It's present on my fork; I've opened a PR to your branch.

QuerthDP commented on scylla/src/client/session.rs:
```diff
@@ -1985,7 +2017,11 @@ where
                 .await
                 .map(|res| res.map_err(RequestError::from))
-                .unwrap_or_else(|_| Err(RequestError::RequestTimeout(timeout))),
+                .unwrap_or_else(|_| {
+                    #[cfg(feature = "metrics")]
+                    self.metrics.inc_request_timeouts();
+                    Err(RequestError::RequestTimeout(timeout))
```
Would you mind to elaborate? I think I didn't understand what's on your mind here.
The wildcard has been just introduced here in commit 55e12acec73bda317029114704606c7bca6665e4.

wprzytula replied:
Yes; it's not you who introduced it. Yet, as we're working close to that spot in code, we could improve it as a bonus :)

I mean replacing `_` with a pattern (e.g., `TimeoutError(_)`), in order to make it explicit what is the type of the parameter that we are ignoring.

QuerthDP commented on docs/source/metrics/metrics.md:
```diff
@@ -29,6 +34,27 @@ println!(
     "99.9 latency percentile: {}",
     metrics.get_latency_percentile_ms(99.9).unwrap()
 );
+
+let snapshot = metrics.get_snapshot().unwrap();
```
That's right, there should be an `?` operator, as well as in other functions using the histogram.

QuerthDP commented on scylla/src/observability/metrics.rs:
```diff
@@ -1,23 +1,196 @@
-    cause: &'static str,
+struct ExponentiallyWeightedMovingAverage {
+    alpha: f64,
+    uncounted: AtomicU64,
```
I copied the names of the fields from CPP-Driver.
I think it was called `uncounted` because it indicates the requests that has been made, but haven't been included in the moving average yet.
Feels quite messy but understandable - anyway we might change the name if you have anything better in mind.

QuerthDP commented:
Changelog:

- rebased @wprzytula's refactor on the branch and squashed his contribution into b3ae5012eb3835cd8a9e386f98946056beb3ed3c
- adjusted EWMA docstrings
- fixed invalid unwraps in examplary code
- added more clarity to the request timeout
- resolved nitpicks

QuerthDP commented:
Changelog:

- rebased on main
- bumped Cargo.lock.msrv as it was changed by other commits

I think this PR should pass ale the CI now.

wprzytula reviewed (approved):
:100: Apart from the open discussion over `PercentileSpeculativeExecutionPolicy`, LGTM!

:tada: You've done a truly great job!

roydahan commented:
@Lorak-mmk do we need your review here in order to merge?

Lorak-mmk commented:
> @Lorak-mmk do we need your review here in order to merge?

I'll be reviewing it soon.

muzarski commented on scylla/src/policies/speculative_execution.rs:
```diff
@@ -79,6 +84,12 @@ impl SpeculativeExecutionPolicy for PercentileSpeculativeExecutionPolicy {
         Duration::from_millis(ms)
     }
+
+    #[cfg(not(feature = "metrics"))]
+    fn retry_interval(&self, _: &Context) -> Duration {
+        warn!("PercentileSpeculativeExecutionPolicy requires the 'metrics' feature to work as intended, defaulting to 100 ms");
+        Duration::from_millis(100)
+    }
```
Sounds reasonable. Current approach is no different from `SimpleSpeculativeExecutionPolicy` with `retry_interval` set to 100ms.

Lorak-mmk replied:
Agreed, `PercentileSpeculativeExecutionPolicy` makes no sense without metrics.

wprzytula commented on scylla/src/observability/metrics.rs:
```diff
@@ -35,6 +35,163 @@ pub struct Snapshot {
+                .store(f64::to_bits(instant_rate), Ordering::Release);
+            self.is_initialized.store(true, Ordering::Release);
+        }
+    }
```
> AFAIU, there can be a logical race if multiple threads execute this method concurrently, correct?

Your observation is correct.



> OTOH, if this is not possible for two threads to enter this method concurrently (or there are some other safety guarantees), I think it should be documented.

Agreed. From what I see in the code, there are safety guarantees in the calling code:
```rust
// Multiple threads could read the same `old_tick`...
let old_tick = self.last_tick.load(ORDER_TYPE);
let new_tick = self.start_time.elapsed().as_nanos() as u64;
let elapsed = new_tick - old_tick;

// _"Problematic"_ `if` - see a comment below.
if elapsed > INTERVAL * 1_000_000_000 {
    let new_interval_start_tick = new_tick - elapsed % (INTERVAL * 1_000_000_000);
    // But then only one will succeed in the following COMPARE EXCHANGE operation.
    if self
        .last_tick
        .compare_exchange(old_tick, new_interval_start_tick, ORDER_TYPE, ORDER_TYPE)
        .is_ok()
    {
        let required_ticks = elapsed / (INTERVAL * 1_000_000_000);
        // So only one thread will do the following ticks.
        // My only concern is that this loop might take so long that another thread
        // enters the _"problematic"_ `if` and then we have a logical race there.
        // BUT this is extremely unlikely, because then the loop would have to take
        // 5 seconds! (INTERVAL * 1e9). So I think we are safe from those races.
        for _ in 0..required_ticks {
            self.one_minute_rate.tick();
            self.five_minute_rate.tick();
            self.fifteen_minute_rate.tick();
        }
    }
}
```
Perhaps my comments could be added to the code, along with a **SHOUTING WARNING** that `tick` must not be called concurrently?

Or, we could use _**advanced type system ~~magic~~ machinery**_ and make `tick` require a `SafetyMark` instance as an argument, which would be constructible only the block guarded by that `COMPARE EXCHANGE` idiom, with the restriction enforced by the visibility rules?

The idea is a bit similar to the `SendAttemptedProof` in `pager.rs`. Actually, I'd happily write such type-system-guaranteed mechanism. WDYT? @muzarski @Lorak-mmk

Lorak-mmk replied:
I did not read the code yet, but if something needs to not have concurrent access, then it should accept `&mut` instead of reinventing it using some type system trickery. Why is that not possible?

wprzytula replied:
Well, it's not possible because of the complex logic involved. From what I described in the comments it's clear that it's not statically provable by the borrow checker that the code is sound.

muzarski commented:
> wdyt @wprzytula @muzarski ?

Yep, sounds good to me.

roydahan commented:
@QuerthDP can you please fix the conflicts and address any comment left? (if any)

QuerthDP commented:
Changelog:

- rebased on main
- added safety comments to the `tick` functions

wprzytula commented on scylla/src/observability/metrics.rs:
```diff
@@ -1,23 +1,204 @@
+    ///
+    /// The first time this function is called, the `rate` is set to the `instant_rate`.
+    /// 
+    /// **WARNING: MUST NOT BE CALLED CONCURRENTLY!**
+    fn tick(&self) {
```
:thought_balloon: This makes a valid case to employ `unsafe`, I believe, to force callers to prove that the contract is upheld.

Lorak-mmk replied:
Will there be a memory corruption if the contract is violated? If not, then it is not a valid use case for unsafe.

wprzytula commented on scylla/src/observability/metrics.rs:
```diff
@@ -1,23 +1,204 @@
+        self.count.load(ORDER_TYPE)
+    }
+
+    fn tick_if_necessary(&self) {
+        // Multiple threads could read the same `old_tick`...
+        let old_tick = self.last_tick.load(ORDER_TYPE);
+        let new_tick = self.start_time.elapsed().as_nanos() as u64;
+        let elapsed = new_tick - old_tick;
+
+        // _"Problematic"_ `if` - see a comment below.
+        if elapsed > INTERVAL * 1_000_000_000 {
+            let new_interval_start_tick = new_tick - elapsed % (INTERVAL * 1_000_000_000);
+            // But then only one will succeed in the following COMPARE EXCHANGE operation.
+            if self
+                .last_tick
+                .compare_exchange(old_tick, new_interval_start_tick, ORDER_TYPE, ORDER_TYPE)
+                .is_ok()
+            {
+                let required_ticks = elapsed / (INTERVAL * 1_000_000_000);
+                // So only one thread will do the following ticks.
+                // My only concern is that this loop might take so long that another thread
+                // enters the _"problematic"_ `if` and then we have a logical race there.
+                // BUT this is extremely unlikely, because then the loop would have to take
+                // 5 seconds! (INTERVAL * 1e9). So I think we are safe from those races.
+                for _ in 0..required_ticks {
+                    self.one_minute_rate.tick();
+                    self.five_minute_rate.tick();
+                    self.fifteen_minute_rate.tick();
+                }
+            }
+        }
+    }
+}
```
:thought_balloon: I believe the body of this function could be put into `unsafe` block in order to warn readers and further implementers, and to allow calling `tick()` after making it an `unsafe` function.

NikodemGapski commented:
Changelog:
- rebased on main
- replaced `LockFreeHistogram` with `AtomicHistogram` (the lock-free version can still be found on [this](https://github.com/QuerthDP/scylla-rust-driver/tree/lock-free-histogram) branch of our fork)
- hid the potential synchronisation metrics errors under `Arc<dyn Error>` (that's where the lock-free errors may go)

QuerthDP commented:
Changelog:

- rebased on main
- added requested `#[non-exhaustive]`

Lorak-mmk commented on scylla/src/network/connection_pool.rs:
```diff
@@ -934,6 +943,17 @@ impl PoolRefiller {
         // As this may may involve resolving a hostname, the whole operation is async.
         let endpoint_fut = self.maybe_translate_for_serverless(endpoint);
 
+        let count_in_metrics = {
+            let metrics = self.metrics.clone();
+            move |connect_result: &Result<_, ConnectionError>| {
+                if connect_result.is_ok() {
```
Nit: prefer `Arc::clone` to `.clone()` to make it explicit that you are not cloning the underlying data structure.

Lorak-mmk commented on scylla/src/observability/metrics.rs:
```diff
@@ -212,6 +212,9 @@ pub struct Metrics {
     retries_num: AtomicU64,
     histogram: Arc<AtomicHistogram>,
     meter: Arc<Meter>,
+    total_connections: AtomicU64,
+    connection_timeouts: AtomicU64,
+    request_timeouts: AtomicU64,
```
Why are those not present in the `Snapshot`? Maybe I don't understand the intention behind the `Snapshot` struct - I thought it is supposed to hold data about the `Metrics` at some point in time?

QuerthDP replied:
The idea about the `Snapshot` struct was indeed to hold data at a certain point of time, but the statistics it contains are only related to the histogram.

I see your point about those 3 metrics, but if we wanted to add them to the `Snapshot`, we could also add the "old" metrics like `errors_num`, which also gives us information about number of errors at some point.
This would make no sense for the whole `Snapshot` to exist, as we would just group every statistic in it.

Furthermore, as you can notice, the `Snapshot` statistics are in fact "captured" in time, while others like `total_connections` are Atomic counters and can be changed whenever the need occurs.
So that "moving" them to the `Snapshot` and saving their current state would also mean that we need some other counter, which is counterproductive.

Lorak-mmk commented on scylla/src/cluster/metadata.rs:
```diff
@@ -452,6 +455,7 @@ impl MetadataReader {
+            #[cfg(feature = "metrics")]
             metrics,
         })
     }
```
@muzarski @wprzytula I'm looking at the commit that puts metrics behind crate future, and I'm wondering:
maybe it is better to restrict those `cfg`'s to the metrics module, by providing alternative implementation which will be active when `cfg(not(feature = "metrics"))`. This implementation would have no fields, but all the same methods, which would just be empty.

Why? The reasons to put metrics behind future are, afaik:
- Performance
- Making histogram dependency optional

Doing what I proposed still achieves those goals, but also remove a lot of conditional compilation flags from other files in the crate, which I think is a great plus for code readability.

Lorak-mmk replied:
The method on `Session` that retrieves the metrics would still be under `#[cfg(feature = "metrics")]` of course.

muzarski replied:
> @muzarski @wprzytula I'm looking at the commit that puts metrics behind crate future, and I'm wondering: maybe it is better to restrict those `cfg`'s to the metrics module, by providing alternative implementation which will be active when `cfg(not(feature = "metrics"))`. This implementation would have no fields, but all the same methods, which would just be empty.

I understand your idea. The only thing I don't get: "but all the same methods, which would just be **empty**". How can a method be empty if it returns something other than `()`? Apart from that, your suggestion sounds reasonable.

Lorak-mmk replied:
This artifical `Metrics` type only needs `inc_*`, `dec_*`, and `log_*` methods because those are used to record data.
All such methods return `()`, apart from `log_query_latency`, which returns `Result`, but keeping the `MetricsError` (maybe as pub(crate) if feature is disabled) is not a problem.
Alternatively it could ignore the error - both call site of this method already do that by using `let _ = ...`.

Lorak-mmk replied:
This sounds like extremely premature optimization. Unless we have proof, or at least reasonable suspicion, that it impacts performance in any significant way, I don't agree to sacrifice readability for optimizations.

wprzytula replied:
Another thing I'm against is having metrics structs, passing them and calling methods on them, even if they are fully non-op. This is nonintuitive and hides the fact that the feature is needed for those operations to really perform meaningful stuff. WDYT @Lorak-mmk ?

Lorak-mmk replied:
That's a much more reasonable objection!
How I see it:
- I believe the runtime impact will not be meaningful, so we should focus on code quality.
- The code that interacts with metric will be in the source anyway. In fact there will be less of it if we remove the cfg lines.
- Connection / Session does not need to care what Metrics will do with the data provided to them. I see at as sort of a compile-time traits. You would not raise the same objections for the code having `Arc<dyn MetricsCollector>` and calling its methods, right?

wprzytula replied:
> * You would not raise the same objections for the code having `Arc<dyn MetricsCollector>` and calling its methods, right?

In fact, I would. The difference is that we have `Option<Arc<dyn Policy>>` instead of `Arc<dyn Policy>` for run-time traits. If you agree to have metrics behind `Option`, then I would not raise this particular objection.

Lorak-mmk replied:
I really don't see the difference in this case. As far as I know we never had any reports about Arc's being a performance problem in the driver. Let's not do micro optimizations unless we have data showing they are useful.
Right now we also always clonde the metrics Arc - as we do with ClusterState Arc.
Please also note, that my proposed change can be reverted if it turns out that performance is a problem here - it will not be a breaking change.

wprzytula replied:
Correct observation, both the change and reverting it are not API-breaking. Therefore, I suggest leaving this as-is and possibly addressing your suggestion after 1.0. WDYT?

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -818,9 +818,6 @@ impl Session {
-        let metrics = Arc::new(Metrics::new());
-
         let pool_config = PoolConfig {
             connection_config,
```
Commit: `metrics: unpub Metrics::new() `

How is this change related to the commit? It removes the duplicate unused creation of `Metrics`.
This problem is not present in the code before this PR, so some other commit must introduce it, right?
In such case such older commit should be edited, instead of making this unrelated change in the older commit.

I suspect that clippy is failing because of such duplicate definition, right? If so, the checklist you filled in the PR (`All commits compile, pass static checks and pass test.`) is false.

Lorak-mmk commented on scylla/src/policies/load_balancing/default.rs:
```diff
@@ -1420,6 +1420,7 @@ mod tests {
                 &HashMap::new(),
+                #[cfg(feature = "metrics")]
                 &Default::default(),
             )
```
This, and few other changes likes this, are present in commit `metrics: unpub Metrics::new() `. Don't they belong to the commit that created the feature?

Lorak-mmk commented on scylla/src/observability/metrics.rs:
```diff
@@ -218,7 +218,7 @@ pub struct Metrics {
-    pub fn new() -> Self {
+    pub(crate) fn new() -> Self {
         Metrics::default()
     }
```
> As the user should not be able to create metrics instance otherwise
> than by `get_metrics()` function, the `Metrics::new()` method shall
> be set to pub(crate) visibility to support only internal usage.

In order to fullfill the goal stated in the commit message, you need to make `impl Default for Metrics` guarded by `cfg(test)`, otherwise user can still create an instance of this struct.

brayniac commented on scylla/src/observability/metrics.rs:
```diff
@@ -120,4 +113,56 @@ impl Metrics {
+    }
+}
+
+impl Default for Metrics {
+    fn default() -> Self {
+        // Config: 64ms error, values in range [0ms, ~262_000ms].
+        // Size: (2^13 + 5 * 2^12) * 8B * 2 ~= 450kB.
+        let grouping_power = 12;
+        let max_value_power = 18;
+
```
FWIW - I think the comment language is confusing and should phrase this in terms of relative error and what the approximate size is for a histogram with whatever final parameters are selected.

I’d personally also consider referencing the calculator we provide so that it’s easy to revisit in the future and understand how these parameters change the range, relative error, and corresponding size of the histogram.

https://observablehq.com/@iopsystems/h2histogram

QuerthDP commented on scylla/src/observability/metrics.rs:
```diff
@@ -37,13 +37,181 @@ pub struct Snapshot {
+
+#[derive(Debug)]
+struct Meter {
+    one_minute_rate: ExponentiallyWeightedMovingAverage,
```
As mentioned before, the implementation comes from cpp-driver.

I can surely agree that the current name is nondescriptive, so I will change it to `RequestRateMeter` as you suggested.

brayniac commented on scylla/src/observability/metrics.rs:
```diff
@@ -133,6 +177,55 @@ impl Metrics {
+        Ok((variance as f64).sqrt() as u64)
+    }
+
+    fn minmax(h: &Histogram) -> Result<(u64, u64), MetricsError> {
+        let mut min = u64::MAX;
+        let mut max = 0;
+        for bucket in h {
+            if bucket.count() == 0 {
+                continue;
+            }
+            let lower_bound = bucket.start();
+            let upper_bound = bucket.end();
+
+            min = u64::min(min, lower_bound);
+            max = u64::max(max, upper_bound);
+        }
+
+        if min > max {
+            Err(MetricsError::Empty)
+        } else {
+            Ok((min, max))
+        }
+    }
```
@Lorak-mmk for this use case, that might not be a bad idea. But when you’re finding the percentiles you’re already doing a full pass through the histogram. I’d suggest finding the min and max to that same pass. I’m hoping it’s being used already but, the percentiles() function allows reading multiple percentiles in a single pass. Min and max can also be found in that pass. I haven’t looked to see where that’s done or why this is done separately, but it is something to consider. 

This code block as-is is likely doing many unnecessary comparisons. Even just finding the first and last non zero indices would be an improvement. I don’t think this needs to compare start against min and end against max for every bucket. 

But, if possible, move this in with wherever the percentiles are read out and this additional pass through the histogram goes away.

QuerthDP commented on scylla/src/observability/metrics.rs:
```diff
@@ -37,13 +37,181 @@ pub struct Snapshot {
+            fifteen_minute_rate: ExponentiallyWeightedMovingAverage::new(
+                1.0 - (-(INTERVAL as f64) / 60.0 / 15.0).exp(),
+            ),
+            count: AtomicU64::new(0),
+            start_time: now,
+            last_tick: AtomicU64::new(now.elapsed().as_nanos() as u64),
+        }
```
The difference maker here is the type.
We're going to modify (concurrently) the `last_tick` while executing other operations so that it is an Atomic variable,
although `start_time` is just a constant, which might be stored as `std::time::Instant`.

mykaul commented:
My >100 comments PR alarm just popped here. Are we on the right track here?

wprzytula commented:
> My >100 comments PR alarm just popped here. Are we on the right track here?

We're just finishing this up.

dkropachev commented:
Closed in favor of https://github.com/scylladb/scylla-rust-driver/pull/1263
